### PR TITLE
Add native_dropout

### PR DESCRIPTION
**Pull request by jjsjann123**
Adds native_dropout to have a reasonable target for torchscript in auto diff. native_dropout has scale and train as arguments in its signature, this makes native_dropout more consistent with other operators and removes conditionals in the autodiff definition.

cc @gmagogsfm

### CODE REVIEW

**Comment by albanD** [comment]
This is a pretty big change for just torchscript support @eellison @suo can you find someone to review this in details?

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -166,15 +166,16 @@
 - func: _debug_has_internal_overlap(Tensor self) -> int
   variants: function
 
-- func: _fused_dropout(Tensor self, float p, Generator? generator=None) -> (Tensor, Tensor)
+- func: native_dropout(Tensor input, float p, float scale, bool train) -> (Tensor, Tensor)
   variants: function
   dispatch:
-    CUDA: fused_dropout_cuda
+    CPU: native_dropout_cpu
+    CUDA: native_dropout_cuda
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:173]: consider making this structured!

Comment (jjsjann123) [reply aten/src/ATen/native/native_functions.yaml:173]: Can you point me to some docs on that? I didn't find it under https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md

I can mimic some porting PRs like #57619.

File: test/backward_compatibility/check_backward_compatibility.py [hunk 2]
```diff
@@ -50,6 +50,8 @@
     ("aten::adaptive_avg_pool3d_backward", datetime.date(9999, 1, 1)),
     ("aten::_embedding_bag_dense_backward", datetime.date(9999, 1, 1)),
     ("aten::randperm", datetime.date(9999, 1, 1)),
+    ("aten::_masked_scale", datetime.date(9999, 1, 1)),
```
Comment (dzhulgakov) [inline test/backward_compatibility/check_backward_compatibility.py:53]: There's no such thing as private API :( it's going to break APEX for example: https://github.com/NVIDIA/apex/search?q=_fused_dropout

perhaps keep the old wrapper for BC?

**Review by dzhulgakov — COMMENTED** [review]
perhaps @ngimel can take a look

File: test/backward_compatibility/check_backward_compatibility.py (hunk 2 above)
Comment (kevinstephano) [reply test/backward_compatibility/check_backward_compatibility.py:53]: We are okay breaking APEX in that case.  That is my code, I can fix it.

**Review by eellison — COMMENTED** [review]
Hey, so I think there are a couple changes we can make to avoid conditionals and also avoid making changes to eager bindings. 

`if use_cuda` -> this conditional was added when cpu fusion was disabled, it's on now, we can remove it.

`if train` there is basically no use case where we want to run the autodiff with `train = False`. We can remove the conditional there and then in [guardDifferentiableGraph](https://github.com/pytorch/pytorch/blob/127c9402d0ae7e3f72aa8ec18dbb9b8f12089bae/torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp#L196) add a check that the `train` argument to `dropout` is `True`. If we really wanted to extend this to full generality, we could profile the `train` argument, preserve the conditional in autodiff, and then add a guard, substitute in the constant, and run constant propagation. I dont think thats really necessary though, but other people feel free to weigh in.

**Comment by ngimel** [comment]
Thanks @eellison, fixing things in symbolic script seems like a much better plan.

**Comment by jjsjann123** [comment]
I agree that doing things in autodiff would be a cleaner way to tackle this.

But I'm growing skeptical of relying on autodiff to do right things lately, though I don't have any real issue in this suggestion :stuck_out_tongue_closed_eyes: 

cc'ing @kevinstephano, do you see any issue with the suggested changes in symbolic_script.cpp? Otherwise, I'll be drafting a PR based on that.

**Comment by jjsjann123** [comment]
Linking relevant issue: #66823

**Comment by jjsjann123** [comment]
Updated per review comments. This PR is ready for review again.

**Comment by jjsjann123** [comment]
Bump for review. Not sure if CI failure is related...

File: aten/src/ATen/native/Dropout.cpp [hunk 3]
```diff
@@ -85,11 +86,37 @@ ALIAS_SPECIALIZATION(_feature_alpha_dropout, true,  true )
 
 } // anomymous namepsace
 
+std::tuple<Tensor,Tensor>
+native_dropout_cpu(const Tensor& input, double p, c10::optional<double> scale, c10::optional<bool> train) {
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply aten/src/ATen/native/Dropout.cpp:90]: This makes sense. Will do

**Comment by jjsjann123** [comment]
Bump for update @ngimel

File: aten/src/ATen/native/cuda/Dropout.cu [hunk 4]
```diff
@@ -343,25 +343,61 @@ fused_dropout_cuda(const Tensor& self, double p, c10::optional<Generator> gen_){
...
-   TORCH_CHECK(mask.scalar_type() == at::ScalarType::Byte, "mask should be torch.uint8 dtype");
+std::tuple<Tensor,Tensor>
+native_dropout_cuda(const Tensor& self, double p, c10::optional<bool> train){
+  // short-cut
+  if (p == 1) {
+    // Technically we don't need to short-cut for inference
+    //
+    // native_dropout_cuda is in derivatives.yaml, so we don't need to add data
+    // dependency from output to input for autograd
+    auto ret = at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Dropout.cu:363]: return should have the same memory format

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -176,6 +176,17 @@
   dispatch:
     CUDA: masked_scale_cuda
 
+- func: native_dropout(Tensor input, float p, bool? train) -> (Tensor, Tensor)
```
Comment (ngimel) [inline aten/src/ATen/native/native_functions.yaml:179]: why is `train` argument optional? It's explicitly specified at all callsites.

File: aten/src/ATen/native/cuda/Dropout.cu [hunk 6]
```diff
@@ -321,14 +319,16 @@ inline void launcher(
 
 } //anonymous namespace
 
+template <typename mask_t>
 std::tuple<Tensor,Tensor>
-fused_dropout_cuda(const Tensor& self, double p, c10::optional<Generator> gen_){
-  auto gen = get_generator_or_default<CUDAGeneratorImpl>(gen_, cuda::detail::getDefaultCUDAGenerator());
-  Tensor ret = at::empty_like(self);
-  Tensor mask = at::empty_like(self, self.options().dtype(kByte));
+dropout_cuda(CUDAGeneratorImpl* gen, const Tensor& self, double p, bool train){
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Dropout.cu:324]: you don't need `train` argument here, non-training case can be handled in `native_dropout_cuda`.

Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Dropout.cu:324]: Sounds fair, will update that.

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (jjsjann123) [reply aten/src/ATen/native/native_functions.yaml:179]: This is trying to make it more like the old `_fused_dropout` API. I think the motivation is to make it an easy drop-replace thing. I could have misunderstood the comment here: https://github.com/pytorch/pytorch/pull/63937#discussion_r738270661

Comment (ngimel) [reply aten/src/ATen/native/native_functions.yaml:179]: Ah I see, but you are not replacing it. Ok, nevermind then.

**Comment by ngimel** [comment]
Lots of internal errors like
```
======================================================================
ERROR: test_mc_dropout_with_loss_attenuation (faim.task.non_temporal.test.uncertain_classifier_test.UncertainClassifierTest)
Test a simple forward pass through the model.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/faim/task/non_temporal/test/non_temporal_mtl_classifier_tests#binary,link-tree/faim/task/non_temporal/test/uncertain_classifier_test.py", line 138, in test_mc_dropout_with_loss_attenuation
    out, fusion_out = scripted_classifier(self.dummy_input)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/faim/task/non_temporal/test/non_temporal_mtl_classifier_tests#binary,link-tree/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "<string>", line 46, in <forward op>
            p1m = 1. - p
            scale = 1. / (float(p1m == 0.) + p1m)
            res,mask = torch.native_dropout(input, p, train)
                       ~~~~~~~~~~~~~~~~~~~~ <--- HERE

            def backward(grad_output):
RuntimeError: result0.storage().use_count() == 1INTERNAL ASSERT FAILED at "buck-out/dev/gen/caffe2/generate-code/autograd/generated/VariableType_2.cpp":7211, please report a bug to PyTorch. function: native_dropout

```

**Comment by jjsjann123** [comment]
> Lots of internal errors like
> 
> ```
> ======================================================================
> ERROR: test_mc_dropout_with_loss_attenuation (faim.task.non_temporal.test.uncertain_classifier_test.UncertainClassifierTest)
> Test a simple forward pass through the model.
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/faim/task/non_temporal/test/non_temporal_mtl_classifier_tests#binary,link-tree/faim/task/non_temporal/test/uncertain_classifier_test.py", line 138, in test_mc_dropout_with_loss_attenuation
>     out, fusion_out = scripted_classifier(self.dummy_input)
>   File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/faim/task/non_temporal/test/non_temporal_mtl_classifier_tests#binary,link-tree/torch/nn/modules/module.py", line 1110, in _call_impl
>     return forward_call(*input, **kwargs)
> RuntimeError: The following operation failed in the TorchScript interpreter.
> Traceback of TorchScript (most recent call last):
> RuntimeError: The following operation failed in the TorchScript interpreter.
> Traceback of TorchScript (most recent call last):
>   File "<string>", line 46, in <forward op>
>             p1m = 1. - p
>             scale = 1. / (float(p1m == 0.) + p1m)
>             res,mask = torch.native_dropout(input, p, train)
>                        ~~~~~~~~~~~~~~~~~~~~ <--- HERE
> 
>             def backward(grad_output):
> RuntimeError: result0.storage().use_count() == 1INTERNAL ASSERT FAILED at "buck-out/dev/gen/caffe2/generate-code/autograd/generated/VariableType_2.cpp":7211, please report a bug to PyTorch. function: native_dropout
> ```

Errr.... I don't quite understand how the error was traced. I thought we are executing a differentiable graph, which should have autograd turned off, but it's still going through autograd dispatch and that's where the error was coming from?!

I think the problem comes from where we are returning `self` in native_dropout_cuda, which is an autograd function. Playing with it also exposed other correctness issues for corner cases. I should have everything patched in the added python tests: https://github.com/pytorch/pytorch/pull/63937/commits/a96311a4aa1c2391d0bd3793cd186eedc52c9935.

I'm trying to repro and verify the fix, which require a debug build that is taking forever.....

**Comment by jjsjann123** [comment]
Verified the issue and fix on debug build :tada: 
It's quite unfortunate that we have to `clone()` inputs. But I'm expecting our freezing workflow would have dropout removed in one of its passes. :crossed_fingers: 

@ngimel Would you mind pulling this in again to double check internal tests?

**Comment by ngimel** [comment]
Ok, internal tests are passing now. I don't know if that change to dropout has real perf implications, but I'm hoping since it's corner cases only, and inference shouldn't go to symbolic_script at all, that should be fine. @eellison thoughts?

**Review by ngimel — APPROVED** [review]

**Comment by helenHlz** [comment]
> Lots of internal errors like
> 
> ```
> ======================================================================
> ERROR: test_mc_dropout_with_loss_attenuation (faim.task.non_temporal.test.uncertain_classifier_test.UncertainClassifierTest)
> Test a simple forward pass through the model.
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/faim/task/non_temporal/test/non_temporal_mtl_classifier_tests#binary,link-tree/faim/task/non_temporal/test/uncertain_classifier_test.py", line 138, in test_mc_dropout_with_loss_attenuation
>     out, fusion_out = scripted_classifier(self.dummy_input)
>   File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/faim/task/non_temporal/test/non_temporal_mtl_classifier_tests#binary,link-tree/torch/nn/modules/module.py", line 1110, in _call_impl
>     return forward_call(*input, **kwargs)
> RuntimeError: The following operation failed in the TorchScript interpreter.
> Traceback of TorchScript (most recent call last):
> RuntimeError: The following operation failed in the TorchScript interpreter.
> Traceback of TorchScript (most recent call last):
>   File "<string>", line 46, in <forward op>
>             p1m = 1. - p
>             scale = 1. / (float(p1m == 0.) + p1m)
>             res,mask = torch.native_dropout(input, p, train)
>                        ~~~~~~~~~~~~~~~~~~~~ <--- HERE
> 
>             def backward(grad_output):
> RuntimeError: result0.storage().use_count() == 1INTERNAL ASSERT FAILED at "buck-out/dev/gen/caffe2/generate-code/autograd/generated/VariableType_2.cpp":7211, please report a bug to PyTorch. function: native_dropout
> ```

I try to use pytorch nvidia ngc container 21.10. The torch version is 1.10.0a0+0aef44c. When I do inference on the scripted model, I got the same error. The official torch 1.10.0 version works fine. Does this error happen when using pytorch backend for inference on triton-21.10 server?

**Comment by jjsjann123** [comment]
> Does this error happen when using pytorch backend for inference on triton-21.10 server?
Likely Triton uses the NGC pytorch container, which could has a buggy version of the implementation. The easiest way would be to update your container so you'll have this PR from upstream merge.

Having said that, if you are using the old container and would prefer patching it instead, here's the commit that fixes the error: https://github.com/pytorch/pytorch/pull/63937/commits/a96311a4aa1c2391d0bd3793cd186eedc52c9935

**Comment by GuWei007** [comment]
> Hey, so I think there are a couple changes we can make to avoid conditionals and also avoid making changes to eager bindings.
> 
> `if use_cuda` -> this conditional was added when cpu fusion was disabled, it's on now, we can remove it.
> 
> `if train` there is basically no use case where we want to run the autodiff with `train = False`. We can remove the conditional there and then in [guardDifferentiableGraph](https://github.com/pytorch/pytorch/blob/127c9402d0ae7e3f72aa8ec18dbb9b8f12089bae/torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp#L196) add a check that the `train` argument to `dropout` is `True`. If we really wanted to extend this to full generality, we could profile the `train` argument, preserve the conditional in autodiff, and then add a guard, substitute in the constant, and run constant propagation. I dont think thats really necessary though, but other people feel free to weigh in.

strongly agree  to del   `use_cuda` and `if_cuda`   in pytorch code as much as possible

**Pull request closed without merging**